### PR TITLE
feat: add runtimeclass control

### PR DESCRIPTION
**Pull request by oliverbaehler**
Signed-off-by: Oliver Baehler <oliver.baehler@hotmail.com>

This PR resolves multiple bugs/features:
- Implements #633 
- Corrects local Development commands in Makefile
- Validate on Image Pod Updates Registry according to tenant configuration

Currently under development, have to test everything tomorrow.

### CODE REVIEW

**Review by prometherion — COMMENTED** [review]
This feature is definitely useful and must be landed!

However, adding a new field to an already published CRD version is not the right way, since it breaks the contract with the previous installation.

I'd ask you to wait for #644 for getting merged and land the `runtimeClass` to the `v1beta2` API group, rather than `v1alpha1`.

And at the same time, maybe, you could decouple the commits regarding optimising the dev environment since these have been already addressed with #650.

**Comment by prometherion** [comment]
@oliverbaehler may I ask you to rebase over the branch `release/v0.2.0` and add this feature to the `v1beta2` API type, please?

File: pkg/webhook/pod/runtimeclass.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+	if len(name) == 0 {
+		return nil, nil
+	}
+
+	obj := &nodev1.RuntimeClass{}
+
+	if err := c.Get(ctx, types.NamespacedName{Name: name}, obj); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
```
Comment (prometherion) [inline pkg/webhook/pod/runtimeclass.go:40]: TIL that a Pod cannot be created using a non-existing Runtime Class: we can drop the `IsNotFound` error check and return it.

File: pkg/webhook/pod/runtimeclass.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+func (h *runtimeClass) OnDelete(client.Client, *admission.Decoder, record.EventRecorder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) *admission.Response {
+		return nil
+	}
+}
+
+func (h *runtimeClass) OnUpdate(client.Client, *admission.Decoder, record.EventRecorder) capsulewebhook.Func {
+	return func(ctx context.Context, req admission.Request) *admission.Response {
+		return nil
+	}
```
Comment (prometherion) [inline pkg/webhook/pod/runtimeclass.go:63]: Can be a pod updated with a newer runtime class?

**Review by prometherion — CHANGES_REQUESTED** [review]

**Review by prometherion — CHANGES_REQUESTED** [review]
Please, copy the `CustomResourceDefinition` for the Tenant resource into the `charts/capsule/crds` folder so we keep it updated and coherent with the changes introduced by this PR.

File: pkg/webhook/pod/runtimeclass.go (hunk 2 above)
Comment (oliverbaehler) [reply pkg/webhook/pod/runtimeclass.go:63]: These are the only field we need to worry about regarding pod updates:

```
spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds`, `spec.tolerations` (only additions to existing tolerat>
```

File: pkg/webhook/pod/runtimeclass.go (hunk 1 above)
Comment (oliverbaehler) [reply pkg/webhook/pod/runtimeclass.go:40]: You are right. will remove:

```
kubectl apply -f sample.yaml 
Error from server (Forbidden): error when creating "sample.yaml": pods "nginx" is forbidden: pod rejected: RuntimeClass "myclass" not found
```

File: pkg/webhook/pod/runtimeclass.go (hunk 2 above)
Comment (prometherion) [reply pkg/webhook/pod/runtimeclass.go:63]: That's great, this simplifies the handling of the webhook, thanks!

File: docs/content/general/tutorial.md [hunk 3]
```diff
@@ -838,6 +838,42 @@ With the said Tenant specification, Alice can create a Pod resource if `spec.pri
...
+kubectl apply -f - << EOF
+apiVersion: capsule.clastix.io/v1beta2
+kind: Tenant
+metadata:
+  name: oil
+spec:
+  owners:
+  - name: alice
+    kind: User
+  RuntimeClasses:
```
Comment (prometherion) [inline docs/content/general/tutorial.md:858]: It must be `runtimeClasses` as the JSON struct tag

File: pkg/webhook/pod/runtimeclass.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+		return &response
+	}
+
+	switch {
+	case allowed == nil:
+		// Enforcement is not in place, skipping it at all
+		return nil
+	case len(runtimeClassName) == 0:
+		// We don't have to force Pod to specify a Priority Class
```
Comment (prometherion) [inline pkg/webhook/pod/runtimeclass.go:97]:
```suggestion
		// We don't have to force Pod to specify a RuntimeClass
```

**Review by prometherion — CHANGES_REQUESTED** [review]

**Review by prometherion — APPROVED** [review]

**Pull request merged**